crates/notary/server/src/config.rs

```diff
@@ -1,46 +1,89 @@
-    pub authorization: AuthorizationProperties,
-    /// The maximum number of concurrent notarization sessions
-    pub concurrency: usize,
+    pub auth: AuthorizationProperties,
 }
 
-impl Default for NotaryServerProperties {
-    fn default() -> Self {
-        Self {
-            server: ServerProperties::default(),
-            notarization: NotarizationProperties::default(),
-            tls: TLSProperties::default(),
-            notary_key: NotarySigningKeyProperties::default(),
-            logging: LoggingProperties::default(),
-            authorization: AuthorizationProperties::default(),
-            concurrency: 32,
+impl NotaryServerProperties {
+    pub fn new(cli_fields: &CliFields) -> Result<Self> {
+        // Uses config file if given.
+        if let Some(config_path) = &cli_fields.config {
+            let mut config: NotaryServerProperties = parse_config_file(config_path)?;
+
+            // Ensures all relative file paths in the config file are prepended with
+            // the config file's parent directory, so that server binary can be run from
+            // anywhere.
+            let parent_dir = Path::new(config_path)
+                .parent()
+                .ok_or(eyre!("Failed to get parent directory of config file"))?
+                .to_str()
+                .ok_or_else(|| eyre!("Failed to convert path to str"))?
+                .to_string();
+
+            // Prepend notarization key paths.
+            if let Some(path) = &config.notarization.private_key_path {
+                config.notarization.private_key_path = Some(prepend_file_path(path, &parent_dir)?);
+            }
+            if let Some(path) = &config.notarization.public_key_path {
+                config.notarization.public_key_path = Some(prepend_file_path(path, &parent_dir)?);
+            }
+            // Prepend TLS key paths.
+            if let Some(path) = &config.tls.private_key_path {
+                config.tls.private_key_path = Some(prepend_file_path(path, &parent_dir)?);
+            }
+            if let Some(path) = &config.tls.certificate_path {
+                config.tls.certificate_path = Some(prepend_file_path(path, &parent_dir)?);
+            }
+            // Prepend auth whitelist path.
+            if let Some(path) = &config.auth.whitelist_path {
+                config.auth.whitelist_path = Some(prepend_file_path(path, &parent_dir)?);
+            }
+
+            Ok(config)
+        } else {
+            let default_config = Config::try_from(&NotaryServerProperties::default())?;
+
+            let config = Config::builder()
+                .add_source(default_config)
+                // Add in settings from environment variables (with a prefix of NS and '_' as
+                // separator).
+                .add_source(
+                    Environment::with_prefix("NS")
+                        .try_parsing(true)
+                        .prefix_separator("_")
+                        .separator("__"),
```

Comment:
are 2 underscores on purpose?

Reply:
yep! it's for nested config like `tls.private_key` whose env var should be `NS_TLS__PRIVATE_KEY`